mruby Rack middleware

i110 opened the pull request:
TODOs:
- [x] request filtering
- [x] ~response fast path~
- [x] ~preserve original request headers~
- [x] performance tuning

kazuho commented on lib/core/subrequest.c:
```diff
@@ -0,0 +1,145 @@
+    return &subreq->super;
+}
+
+void h2o_dispose_subrequest(h2o_req_t *subreq)
```
A function that allocates (or frees) the memory upon initialization should be named "foo_create" or "foo_destroy".
A function that receives a pointer to an object that needs to be initialized (or to be disposed of) should be named "foo_init" or "foo_dispose".

kazuho commented on include/h2o.h:
```diff
@@ -962,6 +962,10 @@ struct st_h2o_req_t {
+        /*
+         * list of request headers
+         */
+        h2o_headers_t headers;
```
Could you please clarify the intent of adding this in this PR?

i110 replied:
Before this PR, the request headers were never modified, so we can always use req->headers for logging (with `%{HEADERNAME}i` notations). But from this PR the users can modify it in mruby like `env['HTTP_USER_AGENT'] = 'another value'`, so I think we should preserve original request headers as other values (scheme, method, path, etc)

kazuho commented on lib/handler/mruby.c:
```diff
@@ -45,6 +45,38 @@
 
 #define FREEZE_STRING(v) MRB_SET_FROZEN_FLAG(mrb_obj_ptr(v))
 
+struct st_subreq_ostream_t {
```
How about changing the struct name to `st_mruby_subreq_t` or something, assuming that the struct is used to hold all information related to the subrequest?

kazuho commented on lib/handler/mruby.c:
```diff
@@ -45,6 +45,38 @@
 #define FREEZE_STRING(v) MRB_SET_FROZEN_FLAG(mrb_obj_ptr(v))
 
+struct st_subreq_ostream_t {
+    h2o_ostream_t super;
```
How about renaming the property to `_ostr_final`? That is the name we use in `struct st_h2o_http1_conn_t` and `h2o_http2_stream_t`.

kazuho commented on include/h2o.h:
```diff
@@ -1322,6 +1332,10 @@ h2o_hostconf_t *h2o_req_setup(h2o_req_t *req);
  */
 void h2o_req_bind_conf(h2o_req_t *req, h2o_hostconf_t *hostconf, h2o_pathconf_t *pathconf);
 
+h2o_subreq_t *h2o_subrequest_create(h2o_req_t *parent, size_t sz);
```
Could we have `h2o_subrequest_init` and `h2o_subrequest_dispose` instead? The reason I ask this is because the approach generally gives the user more freedom in how the object is allocated.

We would be constrained to "create" and "destroy" in some cases (e.g., when the size of the object changes depending on _how_ it is initialized), but we do not have such restriction in this case.

kazuho commented on lib/core/subrequest.c:
```diff
@@ -0,0 +1,128 @@
+    free(subreq);
+}
+
+int h2o_is_subrequest(h2o_req_t *req)
```
Do we need this to be defined as a function? Isn't just doing `req->parent != NULL` sufficient?

kazuho commented on lib/core/subrequest.c:
```diff
@@ -0,0 +1,128 @@
+    return subreq;
+}
+
+void h2o_subrequest_destroy(h2o_subreq_t *subreq)
```
If we agree to change the function to `h2o_subrequest_dispose` (that does not call `free`), I think we can make this an inline function.

kazuho commented on lib/core/subrequest.c:
```diff
@@ -0,0 +1,128 @@
+    subreq->super.processed_at = parent->processed_at;
+
+    return subreq;
+}
```
Assuming that this is the only function to be defined in the source file (see my comments below about other functions), I think we can move the definition to request.c and remove subrequest.c entirely.

kazuho commented on lib/handler/chunked.c:
```diff
@@ -74,6 +74,9 @@ static void on_setup_ostream(h2o_filter_t *self, h2o_req_t *req, h2o_ostream_t *
 {
     chunked_encoder_t *encoder;
 
+    if (h2o_is_subrequest(req))
```
I would appreciate it if you could leave a TODO comment here, stating that http1.c should decide whether chunked encoding should be applied rather than having a separate module to determine if the underlying transport is in fact HTTP/1.1.

kazuho commented on lib/handler/http2_debug_state.c:
```diff
@@ -38,6 +38,9 @@ static int on_req(h2o_handler_t *_self, h2o_req_t *req)
     }
 
     h2o_http2_debug_state_t *debug_state = req->conn->callbacks->get_debug_state(req, self->hpack_enabled);
+    if (debug_state == NULL) {
```
I assume that we have similar code everywhere; one that check if the callback is NULL, if not, calls the callback and check if the return value is NULL. In other words, we are performing the check twice.

I do not think that having such coding pattern is optimal.

Instead, could we require every instance of `http_conn_callbacks_t` to setup all the callback pointers, so that we can call the callback unconditionally? For example, you could define "default" functions that return NULL (e.g. `h2o_http2_debug_state_t *h2o_conn_default_get_debug_state(h2o_req_t *req, int hpack_enabled) { return NULL; }`) and let the callbacks (that do not require a specialized implementation) use the default.

kazuho commented on lib/handler/mruby.c:
```diff
@@ -405,13 +453,14 @@ mrb_value prepare_fibers(h2o_mruby_context_t *ctx)
 
     handler_ctx->handler = handler;
     handler_ctx->shared = get_shared_context(ctx);
+    handler_ctx->pathconf = pathconf;
```
Why does `pathconf` need to be a per loop variable? It is not thread-dependent, and I believe that you can make it a property of `h2o_mruby_handler_t`. The added bonus of making such change will be that you would be able to record the address in `h2o_mruby_register` and you no longer need to change the type of `on_context_init`.

i110 replied:
@kazuho Great idea, thx. I fixed it at https://github.com/h2o/h2o/pull/1217/commits/bc68c17b98b70b27150dc9a46f126b28ef47a797

kazuho commented on lib/handler/mruby/middleware.c:
```diff
@@ -0,0 +1,743 @@
+    h2o_req_t *super = &subreq->super;
+
+    /* create postfilter and final ostream for subreq */
+    h2o_req_filter_t *postfilter = (void *)h2o_add_req_filter(super, sizeof(*postfilter), 1);
```
Do we still need to have postfilter as a new feature now that you have implemented your own sub-request? http1.c and http2/stream.c do not use postfilter and it might be the case that we can do the same for mruby middleware.

kazuho commented on include/h2o.h:
```diff
@@ -1053,23 +1051,27 @@ struct st_h2o_req_t {
+        h2o_buffer_t *buf;
+        void (*cb)(h2o_req_t *req, void *data, h2o_iovec_t error);
+        void *data;
+    } error;
```
Would you mind renaming this to `error_log` or `error_logger`? IMO this is an object that logs the error, not the error itself.

kazuho commented on include/h2o.h:
```diff
@@ -1053,23 +1051,27 @@ struct st_h2o_req_t {
-    H2O_VECTOR(h2o_req_error_log_t) error_logs;
+    struct {
+        h2o_buffer_t *buf;
+        void (*cb)(h2o_req_t *req, void *data, h2o_iovec_t error);
```
The third argument should be named `msg` or something. It's the text (or the message) that describes the error.

kazuho replied:
Do we need both `req` and `data`?
It could be the case that we do not need `data` If we consider the callback as a method belonging to `req`. Or if we consider the callback as a free-standing callback independent to `h2o_req_t`, then `req` should not be an argument.

kazuho commented on lib/core/request.c:
```diff
@@ -593,29 +623,26 @@ void h2o_req_log_error(h2o_req_t *req, const char *module, const char *fmt, ...)
-        H2O_BUILD_ASSERT(sizeof(vecs) / sizeof(vecs[0]) < IOV_MAX);
-        writev(2, vecs, sizeof(vecs) / sizeof(vecs[0]));
+    /* build prefix */
+    char *prefix = h2o_mem_alloc_pool(&req->pool, sizeof("[] in request::") + 32 + errlen), *p = prefix;
```
I would appreciate it if you could adjust the code to avoid introducing a copy of the formatted string (i.e., copying {errbuf, errlen} into `p`).

i110 replied:
fixed at https://github.com/h2o/h2o/pull/1217/commits/391a1e85be27224e59627787a7d9eef2e5daa4c7

kazuho commented on lib/handler/mruby/middleware.c:
```diff
@@ -0,0 +1,722 @@
+        goto Failed; \
+    } \
+    if (stringify) { \
+        v = h2o_mruby_to_str(mrb, v); \
```
Do we need to check if the invocation result in an error (i.e., `mrb->exc` becoming non-null)?

kazuho commented on lib/handler/mruby/middleware.c:
```diff
@@ -0,0 +1,722 @@
+    super->pathconf = ctx->handler->pathconf;
+    super->version = h2o_parse_protocol_version_string(h2o_iovec_init(RSTRING_PTR(server_protocol), RSTRING_LEN(server_protocol)));
+    if (super->version == -1)
+        super->version = 0x101;
```
Do we need to respect `SERVER_PROTOCOL`?

kazuho commented on lib/handler/mruby/middleware.c:
```diff
@@ -0,0 +1,722 @@
+    int gc_arena = mrb_gc_arena_save(mrb);
+    mrb_gc_protect(mrb, env);
+
+    mrb_value dupenv = mrb_funcall(mrb, env, "dup", 0);
```
I would appreciate it if you could do `env = mrb_funcall(..., env, ...)` to replace `env`.
Having both `dupenv` and `env` available as different variables in this big function seems scary to me.

kazuho commented on lib/handler/mruby/middleware.c:
```diff
@@ -0,0 +1,752 @@
+        subreq->conn.super.hosts[1] = NULL;
+    } else {
+        subreq->conn.super.hosts = ctx->handler->pathconf->global->hosts;
+    }
```
Shouldn't the subrequest inherit `h2o_conn_t::hosts` from the parent request?

`hosts` represents the host-level configurations that a request sent over the connection can be bound to.

i110 replied:
addressed at https://github.com/h2o/h2o/pull/1217/commits/4d7e72d8bdf4344bf6f4dfd40f1b18b60e744b02 and https://github.com/h2o/h2o/pull/1217/commits/e9ca27a4fd2d74efc57f0ce9326c71e8fb457a81

NOTE: we changed the behavior of proxy handler to always use globalconf->hosts, so we can  also use it for subreq->conn->hosts because all following handlers use it too

i110 replied:
https://github.com/h2o/h2o/commit/4d7e72d8bdf4344bf6f4dfd40f1b18b60e744b02 
causes a bug that `req->res.content_length` is not cleared when that request is replayed,. https://github.com/h2o/h2o/pull/1217/commits/a01f474811ba1465f99e7d080127659b16dc54b7 fixes that.

kazuho commented:
@i110 What is the reason for prohibiting the modification of `SCRIPT_NAME` in abc42d3?

I assume that it can be done in MRI/Rack and I do not see a reason why we need to prohibit it. IMO our case (a mruby/Rack middleware at path-level) is the same as a MRI/Rack middleware running after a router.

i110 commented:
@kazuho 
> What is the reason for prohibiting the modification of SCRIPT_NAME in abc42d3?

As we discussed the other day, some handlers have prerequisites that req->path matches with req->pathconf->path (for instance: https://github.com/h2o/h2o/blob/master/lib/handler/file.c#L743-L749). `H2O.next.call` is for calling next handlers, so I feel it's natural to prohibit modifying SCRIPT_NAME.

> I assume that it can be done in MRI/Rack

I don't know the middlewares which directly modify `env['SCRIPT_NAME']`, and even if there're I don't think it's good design. In such cases, middlewares should modify PATH_INFO, and modifying SCRIPT_NAME may be the task of Router (like Rack::URLMap). And if the user want to redirect the request internally, there're some specific modules to realize it (Rack::Recursive or my  https://github.com/i110/rack-reprocess :D), which corresponds our `H2O.reprocess.call`.

What do you think?

kazuho commented:
@i110 Thank you for the explanation. I agree.

kazuho commented on lib/core/request.c:
```diff
@@ -221,6 +215,27 @@ static void retain_original_response(h2o_req_t *req)
+{
+    /* use writev(2) to emit error atomically */
+    struct iovec vecs[] = {{prefix.base, prefix.len}, {msg.base, msg.len}, {"\n", 1}};
+    H2O_BUILD_ASSERT(sizeof(vecs) / sizeof(vecs[0]) < IOV_MAX);
```
Should the relational operator be `<=`?

kazuho commented on lib/handler/mruby/middleware.c:
```diff
@@ -0,0 +1,961 @@
+    assert(!mrb_nil_p(subreq->error_stream));
+
+    h2o_iovec_t list[] = {prefix, msg};
+    h2o_iovec_t concat = h2o_concat_list(&subreq->super.pool, list, 2);
```
Please use `h2o_concat(&subreq->super.pool, prefix, msg)` to avoid constructing a temporary object (i.e. `list[]).

kazuho commented on lib/handler/mruby/middleware.c:
```diff
@@ -0,0 +1,961 @@
+
+    /* create subreq */
+    struct st_mruby_subreq_t *subreq = h2o_mem_alloc(sizeof(*subreq));
+    h2o_req_t *super = &subreq->super;
```
Can we refrain from creating an alias? My understanding is that the distance between `subreq` and `&subreq->super` is a constant. Hence we can always use `subreq->super` instead of having `super` that refers to the same object.

kazuho commented on lib/handler/mruby/middleware.c:
```diff
@@ -0,0 +1,961 @@
+#define COND1(str, lit, pos) (COND0(str, lit, pos) && COND0(str, lit, pos + 1) && COND0(str, lit, pos + 2))
+#define COND2(str, lit, pos) (COND1(str, lit, pos) && COND1(str, lit, pos + 3) && COND1(str, lit, pos + 6))
+#define COND(str, lit)       (COND2(str, lit, 0)   && COND2(str, lit, 9)       && COND2(str, lit, 18))
+#define CHECK_KEY(lit) ((sizeof(lit) - 1) == keystr_len && COND(keystr, lit))
```
May I ask the reason for not using `memcmp` here?

I would assume that the C compiler will generate efficient code to compare a prefixed number of bytes. For example, for a CPU that does not have penalty on unaligned load (e.g. x86-64 without cache miss), a compiler might generate 64-bit compare instruction to compare 8 bytes at once.

i110 replied:
This expects compiler optimizations. If we use `memcmp`, comparing a hash key with known keys (e.g. `SCRIPT_NAME`) take the cost in O(nm) where n is the lengthes of hash key and m is sum of the length of known keys. OTOH if we directly write character comparisons like my commit, `-O2` compilation optimizes it by omitting needless comparisons just like using prefix tree, and reduces the cost to O(n). I checked the behaviors of gcc and clang, and took some benchmarks that showed this way is certainly faster than `memcmp`s and other methods (like calculating hashes, etc)

kazuho commented on lib/handler/mruby/middleware.c:
```diff
@@ -0,0 +1,961 @@
+        const mrb_int keystr_len = RSTRING_LEN(key);
+
+        if (CHECK_KEY("CONTENT_LENGTH")) {
+            super->content_length = h2o_strtosize(RSTRING_PTR(value), RSTRING_LEN(value));
```
Do we need an error check here?

kazuho commented on lib/handler/mruby/middleware.c:
```diff
@@ -0,0 +1,961 @@
+        } else if (CHECK_KEY("rack.url_scheme")) {
+            RETRIEVE_ENV(scheme, 1);
+        } else if (keystr_len >= 5 && memcmp(keystr, "HTTP_", 5) == 0) {
+            value = h2o_mruby_to_str(mrb, value);
```
Maybe we need error check after calling `h2o_mruby_to_str`? I see other places that have the same issue.

kazuho commented on lib/handler/mruby/middleware.c:
```diff
@@ -0,0 +1,961 @@
+#define STR_TO_IOVEC(val) h2o_iovec_init(RSTRING_PTR(val), RSTRING_LEN(val))
+
+    /* construct url and parse */
+    h2o_iovec_t *url_comps = alloca(sizeof(*url_comps) * 9);
```
Why not just use `url_comps[9]`? Use of `alloca` is only necessary when the number of elements is unknown at compile-time.

kazuho commented on lib/handler/mruby/middleware.c:
```diff
@@ -0,0 +1,961 @@
+        /* TODO is there any other way to show better error message? */
+        mrb->exc = mrb_obj_ptr(mrb_exc_new_str_lit(mrb, E_ARGUMENT_ERROR, "env variable contains invalid values"));
+        goto Failed;
+    }
```
Could you please elaborate why you need to construct `url_str` then call `h2o_url_parse` here?

Can't you just assign `super->input.scheme`, `super->input.authority`, `super->input.path` using the mruby variables e.g. `scheme`?

i110 replied:
This is for validation. By h2o_url_parse we validate scheme, host, port and path

kazuho commented on lib/handler/mruby/middleware.c:
```diff
@@ -0,0 +1,961 @@
+
+    if (!(mrb_nil_p(server_addr) || mrb_nil_p(server_port))) {
+        subreq->conn.server.host = h2o_strdup(&super->pool, RSTRING_PTR(server_addr), RSTRING_LEN(server_addr));
+        subreq->conn.server.port = h2o_strdup(&super->pool, RSTRING_PTR(server_port), RSTRING_LEN(server_port));
```
Is there a guarantee that both of the variables will be non-nil if either one is?

If not, we have a chance of calling `RSTRING_PTR`, `RSTRING_LEN` against nil. Is that fine?

kazuho commented on lib/handler/mruby/middleware.c:
```diff
@@ -0,0 +1,961 @@
+
+    if (!(mrb_nil_p(remote_addr) || mrb_nil_p(remote_port))) {
+        subreq->conn.remote.host = h2o_strdup(&super->pool, RSTRING_PTR(remote_addr), RSTRING_LEN(remote_addr));
+        subreq->conn.remote.port = h2o_strdup(&super->pool, RSTRING_PTR(remote_port), RSTRING_LEN(remote_port));
```
Same as above.

i110 commented:
@kazuho Thank you for your review. I addressed issues and replied to some of your comments.

BTW: `t/50mruby-error-log.t` sometimes fails, it seems the test cannot read expected log lines. I'm working on it

kazuho commented on lib/handler/mruby.c:
```diff
@@ -289,13 +324,10 @@ static void handle_exception(h2o_mruby_context_t *ctx, h2o_mruby_generator_t *ge
         fprintf(stderr, "mruby raised: %s\n", RSTRING_PTR(mrb_inspect(mrb, mrb_obj_value(mrb->exc))));
     } else {
         assert(generator->req != NULL);
+        assert(generator->req->_generator == NULL);
```
Are we sure that this assertion is always valid?

I believe that a jump from https://github.com/h2o/h2o/blob/af4586f9e7980330bf2b5c2bcf7a967f92eb132d/lib/handler/mruby.c#L1067 has the value set to non-NULL.

i110 replied:
This assertion may be still valid because L1067 shouldn't be called (i.e. if called, it's our programming miss). So, at https://github.com/h2o/h2o/pull/1217/commits/1986bd62065ed603af84bbb524b6b78f75ba95ea,  I removed that code path, as well as making `handle_exception` more robust because it's called from several locations.

i110 commented:
@kazuho Thank you for your review and adding many doc-comments! I addressed the issue you pointed.

kazuho commented:
I am not sure if 1986bd6 addresses the issue pointed out in https://github.com/h2o/h2o/pull/1217#discussion_r182755976.

FWIW, below is the diff between master. If the precondition is that `generator->req` could be NULL, then we should handle the case before calling `h2o_send_error_500`. The same goes for the case when `generator->req->_generator` is NULL.

```
@@ -285,17 +320,12 @@ static void handle_exception(h2o_mruby_context_t *ctx, h2o_mruby_generator_t *ge
     mrb_state *mrb = ctx->shared->mrb;
     assert(mrb->exc != NULL);
 
-    if (generator == NULL) {
+    if (generator == NULL || generator->req->_generator != NULL) {
         fprintf(stderr, "mruby raised: %s\n", RSTRING_PTR(mrb_inspect(mrb, mrb_obj_value(mrb->exc))));
     } else {
-        assert(generator->req != NULL);
         h2o_req_log_error(generator->req, H2O_MRUBY_MODULE_NAME, "mruby raised: %s\n",
                           RSTRING_PTR(mrb_inspect(mrb, mrb_obj_value(mrb->exc))));
-        if (generator->req->_generator == NULL) {
-            h2o_send_error_500(generator->req, "Internal Server Error", "Internal Server Error", 0);
-        } else {
-            h2o_mruby_send_chunked_close(generator);
-        }
+        h2o_send_error_500(generator->req, "Internal Server Error", "Internal Server Error", 0);
     }
     mrb->exc = NULL;
 }
```

i110 commented:
@kazuho 

> If the precondition is that generator->req could be NULL, 

No, `generator->req` is always not NULL if generator is not NULL. OTOH `generator->req->_generator` can be NULL when the response has been started.

i110 commented:
@kazuho 
> Are we fine with the omission?

original [h2o_mruby_send_chunked_close](https://github.com/h2o/h2o/blob/master/lib/handler/mruby/chunked.c#L269-L280) does 2 things:

1. call `body.close`
  - In this PR, it's done in [dispose callback](https://github.com/h2o/h2o/blob/1986bd62065ed603af84bbb524b6b78f75ba95ea/lib/handler/mruby/sender.c#L150) because I couldn't see any reasons to call it here
2. send remaining buffer to the client
  - I thought we don't have to (or shouldn't ?) send remaining buffer in this error case

kazuho commented:
Thank you for the clarification. SGTM.

kazuho commented:
@i110 Please feel free to merge this PR at any time.

i110 commented:
One and a half year! It was so long journey.

proyb6 commented:
I assume the additional feature is ability to modify env?

If it’s similar to x-reproxy, which is more efficientl and when to use which?
